src/contracts/AbstractARM.sol

```diff
@@ -0,0 +1,689 @@
+    }
+
+    function _swapTokensForExactTokens(IERC20 inToken, IERC20 outToken, uint256 amountOut, address to)
+        internal
+        virtual
+        returns (uint256 amountIn)
+    {
+        uint256 price;
+        if (inToken == token0) {
+            require(outToken == token1, "ARM: Invalid out token");
+            price = traderate0;
+        } else if (inToken == token1) {
+            require(outToken == token0, "ARM: Invalid out token");
+            price = traderate1;
+        } else {
+            revert("ARM: Invalid in token");
+        }
+        amountIn = ((amountOut * PRICE_SCALE) / price) + 1; // +1 to always round in our favor
+
+        // Transfer the input tokens from the caller to this ARM contract
+        _transferAssetFrom(address(inToken), msg.sender, address(this), amountIn);
+
+        // Transfer the output tokens to the recipient
+        _transferAsset(address(outToken), to, amountOut);
+    }
+
+    /**
+     * @notice Set exchange rates from an operator account from the ARM's perspective.
+     * If token 0 is WETH and token 1 is stETH, then both prices will be set using the stETH/WETH price.
+     * @param buyT1 The price the ARM buys Token 1 (stETH) from the Trader, denominated in Token 0 (WETH), scaled to 36 decimals.
+     * From the Trader's perspective, this is the sell price.
+     * @param sellT1 The price the ARM sells Token 1 (stETH) to the Trader, denominated in Token 0 (WETH), scaled to 36 decimals.
+     * From the Trader's perspective, this is the buy price.
+     */
+    function setPrices(uint256 buyT1, uint256 sellT1) external onlyOperatorOrOwner {
+        // Limit funds and loss when called by the Operator
+        if (msg.sender == operator) {
+            require(sellT1 >= PRICE_SCALE - MAX_PRICE_DEVIATION, "ARM: sell price too low");
+            require(buyT1 <= PRICE_SCALE + MAX_PRICE_DEVIATION, "ARM: buy price too high");
+        }
+        uint256 _traderate0 = PRICE_SCALE * PRICE_SCALE / sellT1; // base (t0) -> token (t1)
+        uint256 _traderate1 = buyT1; // token (t1) -> base (t0)
+        _setTraderates(_traderate0, _traderate1);
+    }
+
+    function _setTraderates(uint256 _traderate0, uint256 _traderate1) internal {
+        require((PRICE_SCALE * PRICE_SCALE / (_traderate0)) > _traderate1, "ARM: Price cross");
+        traderate0 = _traderate0;
+        traderate1 = _traderate1;
+
+        emit TraderateChanged(_traderate0, _traderate1);
+    }
+
+    ////////////////////////////////////////////////////
+    ///         Liquidity Provider Functions
+    ////////////////////////////////////////////////////
+
+    /// @notice Preview the amount of shares that would be minted for a given amount of assets
+    /// @param assets The amount of liquidity assets to deposit
+    /// @return shares The amount of shares that would be minted
+    function previewDeposit(uint256 assets) external view returns (uint256 shares) {
+        shares = convertToShares(assets);
```

Comment:
⚪️ We should invariant test that `previewDeposit` returns the actual number of shares that are actually minted by `deposit`.

Same redeem.